rfc/0052-sdk3-full-text-search.md

```diff
@@ -566,7 +566,275 @@ A query that matches nothing.
+
+This takes us to a new API, `cluster.search(SearchRequest, SearchOptions)`, which has the flexibility to be able to perform traditional FTS and/or vector queries, and/or any future top-level queries.
+
+The current `cluster.searchQuery()` API will not be deprecated at this time, given the volatility of this new feature.
+But the intent is that the new `cluster.search()` will ultimately be how new users perform any sort of FTS service query, including just a standard FTS `SearchQuery` without vector search.
+
+### API Examples
+All examples will use the reference Java SDK implementation.
+
+1. A traditional FTS query without vector search.  With the existing API:
+
+```java
+SearchQuery query = SearchQuery.matchAll();
+
+scope.searchQuery("search_index_name", query);
+cluster.searchQuery("search_index_name", query);
+```
+
+and with the new API:
+
+```java
+SearchRequest request = SearchRequest.searchQuery(SearchQuery.matchAll());
+
+cluster.search("search_index_name", request);
+scope.search("search_index_name", request);
+```
+
+2. A vector search with a single vector query and no traditional FTS query:
+
+```java
+// Using an imaginary API provided by OpenAI to generate the vector.
+float[] vector = OpenAI.createVectorFor("some query");
+
+SearchRequest request = SearchRequest
+        .vectorSearch(VectorSearch.create(VectorQuery.create("vector_field", vector)));
+
+scope.search("search_index_name", request);
+cluster.search("search_index_name", request);
+```
+
+3. A combined search with both vector search and a traditional FTS query:
+
+```java
+float[] vector = OpenAI.createVectorFor("some query");
+
+SearchRequest request = SearchRequest
+        .searchQuery(SearchQuery.matchAll())
+        .vectorSearch(VectorSearch.create(VectorQuery.create("vector_field", vector)));
+
+scope.search("search_index_name", request);
+cluster.search("search_index_name", request);
+```
+
+4. A more complex example showing a vector search containing multiple vector queries, setting all parameters:
+
+```java
+// Sending multiple `VectorQuery`s, and setting all possible parameters.
+SearchRequest request = SearchRequest
+        .vectorSearch(VectorSearch.create(List.of(
+                VectorQuery.create("vector_field", aVector).numCandidates(2).boost(0.3),
+                VectorQuery.create("vector_field", anotherVector).numCandidates(1).boost(0.7)),
+            vectorSearchOptions().vectorQueryCombination(VectorQueryCombination.AND));
+
+scope.search("search_index_name", request);
+cluster.search("search_index_name", request);
+```
+
+### VectorSearch
+`VectorSearch` construction is platform-idiomatic:
+
+`VectorSearch.create(List<VectorQuery> vectorQueries, [VectorSearchOptions options])`
+
+The first parameter needs to be mandatory.
+
+`VectorSearch.create(VectorQuery vectorQuery, [VectorSearchOptions options])` can be added to make it easier to send a single `VectorQuery`.
```

Comment:
Why does this accept VectorSearchOptions (which only has the operator, which would be unused with a single VectorQuery object)?  Relatedly, in the case of multiple queries, should the operator being used between them not be a more high-level object rather than being part of general options? For instance:
```
VectorSearch.fromMultiple(VectorQueryCombination, List<VectorQuery>)
VectorSearch.create(VectorQuery)
```

Reply:
> Why does this accept VectorSearchOptions (which only has the operator, which would be unused with a single VectorQuery object)

It's forward-looking to when more options will be present in future, but yes, it's not useful right now.  Since discovering what a hard time some SDKs have with adding overloads later, I'm trying to anticipate these situations.
I don't have strong feelings on it - can certainly remove it from this optional sugar.

I'm not a big fan of the `fromMultiple` / VectorQueryCombination idea, I think, for similar reasons - there may well be future options.